### PR TITLE
backend/sdoc/grammar: Add GRAMMAR for REFS-TypeValue variant

### PR DESCRIPTION
**Pull request by GGBeer**
Adding ParentReqReference and FileReference TypeValue-Grammar for GRAMMAR-REFS instead of String.
Note: Adding new ExternalReference and updating the Documentation will be done in a separate step.

### CODE REVIEW

File: strictdoc/backend/sdoc/models/type_system.py [hunk 1]
```diff
@@ -31,55 +31,127 @@ class RequirementFieldType:
...
 class GrammarElementField:
     def __init__(self):
         self.title: str = ""
+        self.gef_type: str = ""
         self.required: bool = False
 
 
 class GrammarElementFieldString(GrammarElementField):
-    def __init__(self, parent, title: str, required: str):
+    def __init__(self, parent, title: str, required: Optional[str]):
```
Comment (stanislaw) [inline strictdoc/backend/sdoc/models/type_system.py:57]: Is there a reason why `required` is set to Optional here? In the grammar, the `REQUIRED` is always provided for a grammar field, so here I would prefer it strict as well.

File: strictdoc/backend/sdoc/error_handling.py [hunk 2]
```diff
@@ -235,6 +236,36 @@ def not_comma_separated_tag_field(  # pylint: disable=too-many-arguments
             filename=filename,
         )
 
+    @staticmethod
+    def invalid_type_value_reference_item(  # pylint: disable=too-many-arguments
```
[the remark this answers is not in the document]
Comment (stanislaw) [reply strictdoc/backend/sdoc/error_handling.py:240]: This folder contains similar cases: `tests/integration/per_document_grammar/validation`. Please add one there.

File: tests/unit/strictdoc/backend/sdoc/test_dsl_passthrough.py [hunk 3]
```diff
@@ -945,3 +955,357 @@ def test_153_grammar_tag():
...
+
+    try:
+        document = reader.read(sdoc_input)
+        assert isinstance(document, Document)
+
+        ll_requirement = document.section_contents[0]
+        references = ll_requirement.references
+        assert len(references) == 0
+
+    except Exception as err:
```
Comment (stanislaw) [inline tests/unit/strictdoc/backend/sdoc/test_dsl_passthrough.py:1158]: Writing the test with try/except, you make the test always pass. If you want to test the positive outcome, remove try/except. If you wish to test that the exception is thrown, rather use this example and write an independent test case for error handling: `test_003_one_range_pragma_begin_req_not_equal_to_end_req`.

**Review by stanislaw — CHANGES_REQUESTED** [review]
Excellent work, let's just polish some edges, and please get the branch to pass all tests.

File: strictdoc/backend/sdoc/error_handling.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (stanislaw) [reply strictdoc/backend/sdoc/error_handling.py:240]: I will add a note to the contributing guideline on how I decide when to add integration or unit or boths types of tests but for now, I would suggest let's add both types.

Let's use this file `test_dsl_grammars.py` for adding a unit test with the validation.

File: strictdoc/backend/sdoc/models/constants.py [hunk 4]
```diff
@@ -24,6 +24,7 @@
     GrammarElementFieldSingleChoice,
     GrammarElementFieldMultipleChoice,
     GrammarElementFieldTag,
+    GrammarElementFieldTypeValue,
```
[the remark this answers is not in the document]
Comment (GGBeer) [reply strictdoc/backend/sdoc/models/constants.py:27]: OK
`GrammarElementFieldReference `=>
```
 [GRAMMAR]
...
  - TITLE: REFS
    TYPE: Reference(ParentReqReference, FileReference)
    REQUIRED: False

```

File: strictdoc/backend/sdoc/error_handling.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (GGBeer) [reply strictdoc/backend/sdoc/error_handling.py:240]: Invalid ReferenceType are catched by the Grammar-Parser.
On using an unlisted ReferenceType item, the 
strictdoc/backend/sdoc/validations/requirement:validate_requirement_field:Line:165..184 raises the StrictDocSemanticError.invalid_reference_type_item Exception.
See UnitTests test_dsl_passthrough.py
def test_163_grammar_refs_file_only()
def test_166_grammar_refs_parent_only():

[the remark this answers is not in the document]
Comment (GGBeer) [reply strictdoc/backend/sdoc/error_handling.py:240]: Getting the iTests to run has given me some headaches. 
Currently using PyCharm 2022..2.4, Python3.10 on Win-Pro-10. 
Had to setup a CentOS Test-VM, but running iTests is still unstable. 
Maybe you can add some hints in the development section.

I will try to add some tests for the ReferenceType variants.

File: strictdoc/backend/sdoc/models/type_system.py (hunk 1 above)
Comment (GGBeer) [reply strictdoc/backend/sdoc/models/type_system.py:57]: Yes, you are right.
Removed the Optional
 `required: Optional[str]` => `required: str`

File: tests/unit/strictdoc/backend/sdoc/test_dsl_passthrough.py (hunk 3 above)
Comment (GGBeer) [reply tests/unit/strictdoc/backend/sdoc/test_dsl_passthrough.py:1158]: Changed the UnitTests test_dsl_passthrough.py
def test_163_grammar_refs_file_only()
def test_166_grammar_refs_parent_only():
into `test that the exception is thrown` variant.

**Comment by GGBeer** [comment]
Still have to rebase the last commits. 
Just wanted to make sure that the CI itest complete without failures.

**Comment by stanislaw** [comment]
@GGBeer,

@mettta and I are merging a big branch with some server and web frontend work just now. This is done to avoid big branches and accumulating conflicts. After I am done with some follow-up work related to that branch, I will get to review this your PR again. Thank you for your patience.

**Comment by stanislaw** [comment]
General comment: I would be curious to learn which problems with itests did you encounter on Windows?

**Comment by GGBeer** [comment]
Must setup a new clean testsystem without any old python, pycharm, pip modules or installed in root vs. user space or whatever which could mess up the itest runs. 
Pytest Unit-Tests run as expected, For iTests I have to execute the lit statement separately.

**Review by stanislaw — APPROVED** [review]
Hereby approved. Please rebase it one last time and I will merge it.

**Review by stanislaw — APPROVED** [review]
Thank you for addressing all review comments!

**Comment by GGBeer** [comment]
Working in a developer mode is fine for me.
I will add some docs into the user guide next. After that I will start a new ExternalReference branch.

Thx,
Guus

**Pull request merged**